### ISSUE TITLE
Weave contacting old hosts-- NOT listed in ipam status. How to fix?

### ISSUE THREAD

**Issue by dcowden** [weaveworks/weave#3401]
Our production cluster is having problems.  We know the issue is somehow related to #3300  and others.

We understand the reclaiming issue, and we know a fix is in progress for that.  

What we've been doing as a workaround is to manually run this:

`weave --local status ipam`

and then remove unreachable peers.

Up until today, that's always worked. But today, we are seeing that even after removing all unreachable peers, we still see weave attempting and connecting to old hosts--even when IPAM shows no unreachable peers.

Here is the output of weave --local status ipam ( note no unreachable peers listed)

```
[root@ip-172-22-21-36 ~]# docker exec $CONTAINER_ID /home/weave/weave --local status ipam
a2:72:ce:d0:74:8c(ip-172-22-21-36.ec2.internal)   131072 IPs (06.2% of total) (20 active)
e2:c3:34:fa:1d:4a(ip-172-22-0-158.ec2.internal)    65536 IPs (03.1% of total) 
0e:30:5d:48:8f:1a(ip-172-22-16-109.ec2.internal)    32768 IPs (01.6% of total) 
fa:93:43:24:ce:ed(ip-172-22-8-124.ec2.internal)    65536 IPs (03.1% of total) 
a6:7f:9c:d4:b7:aa(ip-172-22-5-85.ec2.internal)   262144 IPs (12.5% of total) 
ea:44:8e:b0:8b:f9(ip-172-22-13-48.ec2.internal)   131072 IPs (06.2% of total) 
76:d7:31:af:88:6e(ip-172-22-5-11.ec2.internal)    32768 IPs (01.6% of total) 
b2:8a:72:cb:e1:dc(ip-172-22-13-118.ec2.internal)  1245184 IPs (59.4% of total) 
72:95:b2:4d:0a:54(ip-172-22-21-228.ec2.internal)   131072 IPs (06.2% of total) 

```

But we still see that pod attempting to  connect to old nodes:

```
9/11/18
3:25:58.112 PM	
2018-09-11T19:25:58.112052Z ::  stream=stderr k8s_pod_name=weave-net-h6tjz k8s_pod_id=2e58a5cd-a5bd-11e8-93de-1216123c75f4 k8s_namespace_name=kube-system k8s_host=ip-172-22-21-36.ec2.internal k8s_controller-revision-hash=715543512 k8s_name=weave-net k8s_pod-template-generation=4 k8s_role_kubernetes_io/networking=1  INFO: 2018/09/11 19:25:58.111709 ->[172.22.13.172:6783] error during connection attempt: dial tcp4 :0->172.22.13.172:6783: connect: connection timed out
env =	prod host =	ip-172-22-21-36 source =	/apps/fluent-out/kubernetes.log.20180911.log sourcetype =	container_logs timeendpos =	27
9/11/18
3:25:55.167 PM	
2018-09-11T19:25:55.167990Z ::  stream=stderr k8s_pod_name=weave-net-h6tjz k8s_pod_id=2e58a5cd-a5bd-11e8-93de-1216123c75f4 k8s_namespace_name=kube-system k8s_host=ip-172-22-21-36.ec2.internal k8s_controller-revision-hash=715543512 k8s_name=weave-net k8s_pod-template-generation=4 k8s_role_kubernetes_io/networking=1  INFO: 2018/09/11 19:25:55.167684 ->[172.22.8.127:6783] error during connection attempt: dial tcp4 :0->172.22.8.127:6783: connect: connection timed out
env =	prod host =	ip-172-22-21-36 source =	/apps/fluent-out/kubernetes.log.20180911.log sourcetype =	container_logs timeendpos =	27
9/11/18
3:25:20.479 PM	
2018-09-11T19:25:20.479820Z ::  stream=stderr k8s_pod_name=weave-net-h6tjz k8s_pod_id=2e58a5cd-a5bd-11e8-93de-1216123c75f4 k8s_namespace_name=kube-system k8s_host=ip-172-22-21-36.ec2.internal k8s_controller-revision-hash=715543512 k8s_name=weave-net k8s_pod-template-generation=4 k8s_role_kubernetes_io/networking=1  INFO: 2018/09/11 19:25:20.479638 ->[172.22.16.75:6783] error during connection attempt: dial tcp4 :0->172.22.16.75:6783: connect: connection timed out
env =	prod host =	ip-172-22-21-36 source =	/apps/fluent-out/kubernetes.log.20180911.log sourcetype =	container_logs timeendpos =	27
9/11/18
3:24:38.751 PM	
2018-09-11T19:24:38.751900Z ::  stream=stderr k8s_pod_name=weave-net-h6tjz k8s_pod_id=2e58a5cd-a5bd-11e8-93de-1216123c75f4 k8s_namespace_name=kube-system k8s_host=ip-172-22-21-36.ec2.internal k8s_controller-revision-hash=715543512 k8s_name=weave-net k8s_pod-template-generation=4 k8s_role_kubernetes_io/networking=1  INFO: 2018/09/11 19:24:38.751631 ->[172.22.5.159:6783] error during connection attempt: dial tcp4 :0->172.22.5.159:6783: connect: connection timed out
env =	prod host =	ip-172-22-21-36 source =	/apps/fluent-out/kubernetes.log.20180911.log sourcetype =	container_logs timeendpos =	27
9/11/18
3:22:58.271 PM	
2018-09-11T19:22:58.271834Z ::  stream=stderr k8s_pod_name=weave-net-h6tjz k8s_pod_id=2e58a5cd-a5bd-11e8-93de-1216123c75f4 k8s_namespace_name=kube-system k8s_host=ip-172-22-21-36.ec2.internal k8s_controller-revision-hash=715543512 k8s_name=weave-net k8s_pod-template-generation=4 k8s_role_kubernetes_io/networking=1  INFO: 2018/09/11 19:22:58.271652 ->[172.22.13.22:6783] error during connection attempt: dial tcp4 :0->172.22.13.22:6783: connect: connection timed out

```

I had noticed this but not put too much priority on it, since the underlying issue is in-process. This turned out to be a mistake.  Today in production, one node in a similar state as above became completely unresponsive and unable to run workloads. These messages were in the kubelet logs:

```
Sep 11 15:00:48 ip-172-22-5-85.colinx.com kubelet[26696]: E0911 15:00:48.027998   26696 kuberuntime_manager.go:646] createPodSandbox for pod "fluentd-log-collector-7rj88_ops(873bc81e-b5e8-11e8-a439-12dbc5d74712)" failed: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "fluentd-log-collector-7rj88_ops" network: unable to allocate IP address: Post http://127.0.0.1:6784/ip/043abf8a8261ea4bd530395b71ab8273374ae7a4650efe575188707cd925e55e: dial tcp 127.0.0.1:6784: connect: connection refused
Sep 11 15:00:48 ip-172-22-5-85.colinx.com kubelet[26696]: E0911 15:00:48.028056   26696 pod_workers.go:186] Error syncing pod 873bc81e-b5e8-11e8-a439-12dbc5d74712 ("fluentd-log-collector-7rj88_ops(873bc81e-b5e8-11e8-a439-12dbc5d74712)"), skipping: failed to "CreatePodSandbox" for "fluentd-log-collector-7rj88_ops(873bc81e-b5e8-11e8-a439-12dbc5d74712)" with CreatePodSandboxError: "CreatePodSandbox for pod \"fluentd-log-collector-7rj88_ops(873bc81e-b5e8-11e8-a439-12dbc5d74712)\" failed: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod \"fluentd-log-collector-7rj88_ops\" network: unable to allocate IP address: Post 
```

Again, we understand that a fix for the root issue (not cleaning up old peers ) is underway.  What we need now is a way to clean up old peers that works even when there are no unreachable peers in ipam status, and that doesn't involve stopping all the workloads.

Our cluster is k8s 1.10.3, based on kops 1.10.beta in AWS

**Comment by murali-reddy** [comment]
@dcowden thanks for reporting this issue

>  one node in a similar state as above became completely unresponsive and unable to run workloads. 

Based on the kubelet log `NetworkPlugin cni failed to set up pod` I assume it is weave-net pods that is not responding or failed to setup pod. When in this state did you see any activity/errors in the weave pod logs?

**Comment by dcowden** [comment]
I'm going to pull all of the logs from all weave pods during the event. the file will be large, but it will have all we can offer in terms of diagnosis.

One message I see that's really odd is this one:

`WARN: 2018/09/11 20:06:05.876256 Vetoed installation of hairpin flow FlowSpec{keys: [EthernetFlowKey{src: ea:44:8e:b0:8b:f9, dst: aa:d8:e1:9c:db:8a} InPortFlowKey{vport: 1}], actions: [OutputAction{vport: 1}]}`

**Comment by dcowden** [comment]
It looks like the first unusual message I see in the logs is at 1:15pm. each weave pod reports a message like this one:

`
2018-09-11T17:15:12.611382Z ::  stream=stderr k8s_pod_name=weave-net-bg27v k8s_pod_id=39f582d5-a64c-11e8-93de-1216123c75f4 k8s_namespace_name=kube-system k8s_host=ip-172-22-5-11.ec2.internal k8s_controller-revision-hash=715543512 k8s_name=weave-net k8s_pod-template-generation=4 k8s_role_kubernetes_io/networking=1  DEBU: 2018/09/11 17:15:12.610249 EVENT UpdatePod {"metadata":{"annotations":{"ad.datadoghq.com/ptplace-bff.check_names":"[\"prometheus\"]","ad.datadoghq.com/ptplace-bff.init_configs":"[{}]","ad.datadoghq.com/ptplace-bff.instances":"[ { \"prometheus_url\": \"http://%%host%%:8080/metrics/\", \"namespace\": \"ptplace-bff\", \"metrics\": [ \"*\" ] } ]"},"creationTimestamp":"2018-08-30T19:23:57Z","generateName":"ptplace-bff-55698bd6d7-","labels":{"affiliation":"colinx","app":"ptplace","billing":"ecom","component":"ptplace-bff","dr":"mission-critical","heritage":"Tiller","pod-template-hash":"1125468283","release":"ptplace-bff"},"name":"ptplace-bff-55698bd6d7-rw94q","namespace":"default","resourceVersion":"42878536","selfLink":"/api/v1/namespaces/default/pods/ptplace-bff-55698bd6d7-rw94q","uid":"3dbfc523-ac8a-11e8-a439-12dbc5d74712"},"spec":{"containers":[{"image":"quay.io/colinx/ptplace:1.0.9430","imagePullPolicy":"Always","lifecycle":{"preStop":{"exec":{"command":["/home/appuser/preStop.sh"]}}},"name":"ptplace-bff","ports":[{"containerPort":8080,"protocol":"TCP"}],"readinessProbe":{"failureThreshold":3,"httpGet":{"path":"/ptpr/api/heartbeat","port":8080,"scheme":"HTTP"},"initialDelaySeconds":90,"periodSeconds":5,"successThreshold":1,"timeoutSeconds":90},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File"}],"dnsPolicy":"ClusterFirst","imagePullSecrets":[{"name":"gitlab"},{"name":"quay"}],"nodeName":"ip-172-22-5-85.ec2.internal","restartPolicy":"Always","schedulerName":"default-scheduler","securityContext":{},"serviceAccount":"colinx","serviceAccountName":"colinx","terminationGracePeriodSeconds":43200},"status":{"conditions":[{"lastProbeTime":null,"lastTransitionTime":"2018-08-30T19:23:57Z","status":"True","type":"Initialized"},{"lastProbeTime":null,"lastTransitionTime":"2018-08-30T19:25:48Z","status":"True","type":"Ready"},{"lastProbeTime":null,"lastTransitionTime":"2018-08-30T19:23:57Z","status":"True","type":"PodScheduled"}],"hostIP":"172.22.5.85","phase":"Running","podIP":"100.123.0.5","qosClass":"Guaranteed","startTime":"2018-08-30T19:23:57Z"}} {"metadata":{"annotations":{"ad.datadoghq.com/ptplace-bff.check_names":"[\"prometheus\"]","ad.datadoghq.com/ptplace-bff.init_configs":"[{}]","ad.datadoghq.com/ptplace-bff.instances":"[ { \"prometheus_url\": \"http://%%host%%:8080/metrics/\", \"namespace\": \"ptplace-bff\", \"metrics\": [ \"*\" ] } ]"},"creationTimestamp":"2018-08-30T19:23:57Z","generateName":"ptplace-bff-55698bd6d7-","labels":{"affiliation":"colinx","app":"ptplace","billing":"ecom","component":"ptplace-bff","dr":"mission-critical","heritage":"Tiller","pod-template-hash":"1125468283","release":"ptplace-bff"},"name":"ptplace-bff-55698bd6d7-rw94q","namespace":"default","resourceVersion":"45031531","selfLink":"/api/v1/namespaces/default/pods/ptplace-bff-55698bd6d7-rw94q","uid":"3dbfc523-ac8a-11e8-a439-12dbc5d74712"},"spec":{"containers":[{"image":"quay.io/colinx/ptplace:1.0.9430","imagePullPolicy":"Always","lifecycle":{"preStop":{"exec":{"command":["/home/appuser/preStop.sh"]}}},"name":"ptplace-bff","ports":[{"containerPort":8080,"protocol":"TCP"}],"readinessProbe":{"failureThreshold":3,"httpGet":{"path":"/ptpr/api/heartbeat","port":8080,"scheme":"HTTP"},"initialDelaySeconds":90,"periodSeconds":5,"successThreshold":1,"timeoutSeconds":90},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File"}],"dnsPolicy":"ClusterFirst","imagePullSecrets":[{"name":"gitlab"},{"name":"quay"}],"nodeName":"ip-172-22-5-85.ec2.internal","restartPolicy":"Always","schedulerName":"default-scheduler","securityContext":{},"serviceAccount":"colinx","serviceAccountName":"colinx","terminationGracePeriodSeconds":43200},"status":{"message":"The node was low on resource: nodefs.","phase":"Failed","reason":"Evicted","startTime":"2018-08-30T19:23:57Z"}}`

**Comment by murali-reddy** [comment]
@dcowden thanks for sharing the logs. 

Issue you reported which is `error during connection attempt: dial tcp4 :0->172.22.13.22:6783: connect: connection timed out` is basically noise in the logs which should not have any functional impact. Like you mentioned the PR in progress will fix this issue. But very unlikely this can be cause for node unresponsiveness. 

Looking at the logs there are errors which are indicating nodes were low on resources.

For e.g
```
status":{"message":"The node was low on resource: nodefs.","phase":"Failed","reason":"Evicted","startTime":"2018-08-30T19:23:57Z"}}
```
```
"status":{"message":"Pod The  node was low on resource: [DiskPressure].","phase":"Failed","reason":"Evicted","startTime":"2018-09-11T17:05:14Z"}}
```

Are you sure there are no other errors/alerts in the cluster other than errors in the Weave pods?

Also is it possible that logs in the Weave pods were so noisy that nodes ran out of disk space?

**Comment by dcowden** [comment]
@murali-reddy Yes, i think you're right. we've independently concluded that the root cause of this issue was disk pressure ( nodefs). I think that's the root issue of our outage.

That said, we're still a little concerned about the weave logs showing connecting to hosts that are old, even when they are not listed in ipam status.

**Comment by murali-reddy** [comment]
> That said, we're still a little concerned about the weave logs showing connecting to hosts that are old, even when they are not listed in ipam status.

@dcowden If you don't do rmpeer if would see these logs, but are you seeing that, there is attempt by weave to connect to the hosts that are deleted and were removed from weave peer list with rmpeer?

I can not observe this behaviour. Once a peer is removed with  rmpeer there is no attempt to connect to the deleted peer.

**Comment by bboreham** [comment]
"attempt to connect" is totally separate from IPAM.  There is a `weave forget` command for the latter, distinct from `weave rmpeer`.

(yes, it's confusing)

**Comment by murali-reddy** [comment]
@bboreham thanks. So it seems logical to do `weave forget` along with `weave rmpeer` in #3399?

**Comment by bboreham** [comment]
Yes.

**Comment by rade** [comment]
But isn't it the case that a) `weave forget` is a local operation, rather than cluster-wide, and b) we do not know the relationship between peer names and peer addresses?

**Comment by bboreham** [comment]
a) #3399 runs everywhere; yes you have to do this outside of the only-run-once logic
b) that's a harder problem.  better to just replace the entire list, i.e. `weave connect --replace`

**Comment by murali-reddy** [comment]
This symptoms reported in this issue are fixed in 2.5.0 [release](https://github.com/weaveworks/weave/releases) by  #3372, #3399. Please reopen if you still see the issue with 2.5.0 release.
